engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/MergedDataIndex.java

```diff
@@ -184,27 +331,59 @@ private static Table loadIndexTableAndShiftRowSets(
                     String.join(", ", keyColumnNames), location));
         }
         final Table indexTable = dataIndex.table();
-        return indexTable.coalesce().update(List.of(SelectColumn.ofStateless(new FunctionalColumn<>(
-                dataIndex.rowSetColumnName(), RowSet.class,
-                ROW_SET_COLUMN_NAME, RowSet.class,
-                (final RowSet rowSet) -> rowSet
-                        .shift(RegionedColumnSource.getFirstRowKey(Math.toIntExact(locationRowKey)))))));
+        final long shiftAmount = RegionedColumnSource.getFirstRowKey(Math.toIntExact(locationRowKey));
+        final Table coalesced = indexTable.coalesce();
+
+        // pull the key columns into memory while we are parallel;
+        final Table withInMemorykeyColumns = coalesced.update(keyColumnNames);
+
+        final Selectable shiftFunction;
+        if (shiftAmount == 0) {
+            shiftFunction =
+                    Selectable.of(ColumnName.of(ROW_SET_COLUMN_NAME), ColumnName.of(dataIndex.rowSetColumnName()));
+        } else {
+            shiftFunction = new FunctionalColumn<>(
+                    dataIndex.rowSetColumnName(), RowSet.class,
+                    ROW_SET_COLUMN_NAME, RowSet.class,
+                    (final RowSet rowSet) -> rowSet.shift(shiftAmount));
+        }
+        // the rowset column shift need not occur until we perform the rowset merge operation - which is either
+        // lazy or part of an update [which itself can be parallel].
+        return withInMemorykeyColumns.updateView(List.of(SelectColumn.ofStateless(shiftFunction)));
     }
 
-    private RowSet mergeRowSets(
+    private static RowSet mergeRowSets(
             @SuppressWarnings("unused") final long unusedRowKey,
             @NotNull final ObjectVector<RowSet> keyRowSets) {
+        final long numRowSets = keyRowSets.size();
+
+        if (numRowSets == 1) {
+            // we steal the reference, the input is never used again
+            return keyRowSets.get(0);
+        }
         final RowSetBuilderSequential builder = RowSetFactory.builderSequential();
-        try (final CloseableIterator<RowSet> rowSets = keyRowSets.iterator()) {
-            rowSets.forEachRemaining(builder::appendRowSequence);
+
+        if (USE_PARALLEL_LAZY_FETCH) {
+            LongStream.range(0, numRowSets).parallel().mapToObj(keyRowSets::get)
+                    .sorted(Comparator.comparingLong(RowSet::firstRowKey)).forEachOrdered(rs -> {
+                        builder.appendRowSequence(rs);
+                        rs.close();
+                    });
+        } else {
+            try (final CloseableIterator<RowSet> rowSets = keyRowSets.iterator()) {
+                rowSets.forEachRemaining(rs -> {
```

Comment:
You want to restore this version if I have talked you into `update` for shifted row set loading when not building a lazy table.